lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -205,6 +205,155 @@ static Node *simplifyNode(Node *node, Function *F) {
   return node;
 }
 
+// 2D vector of dimensions mapping a dimension that is reshaped into either 1
+// or 2 dimensions by the first reshape node in a ChannelShuffle operation
+using ChannelShuffleDimGroups =
+    llvm::SmallVector<llvm::SmallVector<unsigned_t, 2>, max_tensor_dimensions>;
+
+/// fills \p dimGroups with the groups of dimensions of the result of \p node
+/// that get separated. This grouping is only possible if \p node is the first
+/// Reshape node in a ChannelShuffle operation (Reshape->Tranpose->Reshape)
+/// \returns true if successfully grouped output dimensions and false otherwise
+static bool
+groupDimensionsFromChannelShuffleReshape(const ReshapeNode &node,
+                                         ChannelShuffleDimGroups &dimGroups) {
+  const auto dims = node.getDims();
```

Comment:
I think it's better to use concrete types(instead of auto) when the return type of the RHS is not clear.
This is of course arguable but it makes code easier to read and review.